Introduce ChunkShardingSpec as a model sharding specification.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58517 ShardedTensor framework for ChunkedShardingSpec
* #57409 Add GenericShardingSpec for generic tensor sharding.
* **#55728 Introduce ChunkShardingSpec as a model sharding specification.**

Full design: https://github.com/pytorch/pytorch/issues/55207

This PR introduces ChunkShardingSpec (SingleShardingSpec in the design). Used
the name ChunkShardingSpec since it is very similar to `torch.chunk` in terms
of how a Tensor is split up and feels more clear compared to SingleShardingSpec.

Differential Revision: [D27694108](https://our.internmc.facebook.com/intern/diff/D27694108/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27694108/)!

pritamdamania committed:
Introduce ChunkShardingSpec as a model sharding specification.

Full design: https://github.com/pytorch/pytorch/issues/55207

This PR introduces ChunkShardingSpec (SingleShardingSpec in the design). Used
the name ChunkShardingSpec since it is very similar to `torch.chunk` in terms
of how a Tensor is split up and feels more clear compared to SingleShardingSpec.

Differential Revision: [D27694108](https://our.internmc.facebook.com/intern/diff/D27694108/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27694108/)!

[ghstack-poisoned]

SciPioneer commented on torch/distributed/utils.py:
```diff
@@ -0,0 +1,65 @@
+import torch
+
+def _parse_remote_device(remote_device: str):
```
I understand that compared with `_parse_remote_device` in `torch/distributed/rpc/utils.py`, here we also want to support another format `rank:<rank>/<device>`. Shall we unify these two very similar functions?

SciPioneer commented on torch/distributed/utils.py:
```diff
@@ -0,0 +1,65 @@
+                1. "<workername>/<device>", where the device field can be parsed as torch.device type.
+                   E.g., "trainer0/cpu", "trainer0", "ps0/cuda:0".
+                   In addition, the device field can be optional and the default value is "cpu".
+                2. "rank:<rank>/<device>", where <rank> is the rank of the
```
Thanks for unifying the util functions for both sharded tensor and remote module! This means that wow we are also supporting the second format for `RemoteModule`. Is there any documentation that we should update for `RemoteModule`?

SciPioneer replied:
It may be even better to just reuse `DevicePlacement` class in `RemoteModule`, then we don't need to maintain two copies of valid formats.

SciPioneer commented on torch/distributed/_sharding_spec/api.py:
```diff
@@ -0,0 +1,98 @@
+            (ex: ["rank:0/cuda:0", "rank:1/cuda:1"]) or a list of custom
+            placement specs.
+
+            The device can be a local device or a remote device specified by one
```
Not against this functionality for supporting sharding a tensor across remote devices. Out of curiosity, have you seen any performance gain by such across-node tensor sharding? It seems that this will increase a lot of communication overheads.

SciPioneer commented on torch/distributed/_sharding_spec/_internals.py:
```diff
@@ -0,0 +1,22 @@
+
+def is_valid_device(device):
+    """
+    Checks if this is a valid remote/local device.
```
Ni: Change the order of "remote/local" here to be consistent with the implementation?

SciPioneer commented on torch/distributed/_sharding_spec/api.py:
```diff
@@ -0,0 +1,98 @@
+                raise ValueError(f'{dev} is not a valid device')
+
+    @staticmethod
+    def _verify_dim(dim):
```
Shall we also very that the value of `dim` is in the range of the tensor's dimensionality? It cannot be negative or a very large value.

SciPioneer commented on test/distributed/_sharding_spec/test_sharding_spec.py:
```diff
@@ -0,0 +1,54 @@
+
+class TestShardingSpec(TestCase):
+
+    def test_device_placement(self):
```
Since you have already reused `_parse_remote_device` util function, it will be great to also share the duplicate test cases for validation (probably by sharing two const lists to represent valid and invalid inputs). We may even consider reusing `DevicePlacement` in `RemoteModule` class, as `DevicePlacement` is more generic than tensor sharding.

rohan-varma commented on torch/distributed/utils.py:
```diff
@@ -0,0 +1,65 @@
+    elif len(fields) > 2:
+        raise ValueError(
+            f"Could not parse remote_device: {remote_device}. The valid format is "
+            "'<workername>/<device>' or 'rank:<rank>/<device>"
```
nit: dedupe these error messages into a constant str?

pritamdamania committed:
Update on "Introduce ChunkShardingSpec as a model sharding specification."


Full design: https://github.com/pytorch/pytorch/issues/55207

This PR introduces ChunkShardingSpec (SingleShardingSpec in the design). Used
the name ChunkShardingSpec since it is very similar to `torch.chunk` in terms
of how a Tensor is split up and feels more clear compared to SingleShardingSpec.

Differential Revision: [D27694108](https://our.internmc.facebook.com/intern/diff/D27694108/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27694108/)!

[ghstack-poisoned]

SciPioneer commented on test/distributed/_sharding_spec/test_sharding_spec.py:
```diff
@@ -0,0 +1,54 @@
+        with self.assertRaisesRegex(ValueError, "not a valid device"):
+            DevicePlacement("foo:0")
+        with self.assertRaisesRegex(ValueError, "not a valid device"):
+            DevicePlacement("rank:0/cuda:foo")
```
Can we try a few more test cases?
1) "cuda:100"
2) "cpu2"

SciPioneer commented:
Please fix the test failures.

divchenko commented on torch/distributed/_sharding_spec/api.py:
```diff
@@ -0,0 +1,98 @@
+        """
+        return self._device
+
+class ChunkShardingSpec(PlacementSpec):
```
ChunkShardingSpec , DevicePlacement and PlacementSpec - naming seems a bit off. Some have 'Placement', some have 'Spec' in their names. It's not obvious they belong to the same class hierarchy.

pritamdamania committed:
Update on "Introduce ChunkShardingSpec as a model sharding specification."


Full design: https://github.com/pytorch/pytorch/issues/55207

This PR introduces ChunkShardingSpec (SingleShardingSpec in the design). Used
the name ChunkShardingSpec since it is very similar to `torch.chunk` in terms
of how a Tensor is split up and feels more clear compared to SingleShardingSpec.

Differential Revision: [D27694108](https://our.internmc.facebook.com/intern/diff/D27694108/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27694108/)!

[ghstack-poisoned]

mrshenli commented on torch/distributed/utils.py:
```diff
@@ -0,0 +1,56 @@
+    )
+
+    fields = remote_device.split("/")
+    if len(fields) == 2:
```
Looks like this method only checks whether the format is correct. Do we also need to check whether the ranks are valid before it is used by, e.g., DevicePlacementSpec?

pritamdamania87 replied:
The specification itself might be tied to different runtime environments. For example, you can store this specification with a checkpoint and then load it during checkpoint load time. As a result, I thought its better not to validate runtime environment here. For example, we don't verify GPU indexes either.

mrshenli replied:
I see. That means, we consider the sharding spec as counterparts of Device/RemoteDevice, and the check will be enforced at the loading time?

pritamdamania87 replied:
Right that is what I had in mind.

SciPioneer replied:
Agree. This syntax-only non-semantics check is consistent with the remote device validation in RPC framework.

mrshenli commented on torch/distributed/utils.py:
```diff
@@ -0,0 +1,56 @@
+    if len(fields) == 2:
+        [on, device] = fields
+    elif len(fields) == 1:
+        on = fields[0]
```
ditto for checking the existence of worker names

pritamdamania87 replied:
same as above, I think this would be done at a higher level like when you actually create a ShardedTensor

mrshenli commented on torch/distributed/utils.py:
```diff
@@ -0,0 +1,56 @@
+    else:
+        raise ValueError(PARSE_ERROR)
+
+    # Since the workername in the input remote device won't be validated until the created remote module is executed,
```
This means we expect ShardingSpec to be created before `init_process_group` and `init_rpc`? If so, does it mean users need to do another round of correctness verification after c10d/rpc are initialized? Will it make sense to require that ShardingSpec can only be created after c10d/rpc initialization, so that we can do more correctness checks here?

pritamdamania87 replied:
I don't think ShardingSpec creation is tied to `init_process_group` or `init_rpc` and it shouldn't be tied to them. The specification should be independent of the runtime environment such that I can apply the same specification to different training jobs as an example.

SciPioneer commented on torch/distributed/_sharding_spec/api.py:
```diff
@@ -0,0 +1,98 @@
+            The dimension to shard on, could be an integer representing the
+            dimension or a string in case of named tensors where dimensions are
+            named.
+        placement(List[Device] or List[PlacementSpec]):
```
Should `List[PlacementSpec]` be `List[DevicePlacementSpec]`? Otherwise, it means that a `ChunkShardingSpec` can be represented by `List[ChunkShardingSpec]`, which means that the sharding here can be hierarchical.

pritamdamania87 replied:
I kept it `List[PlacementSpec]` for extra flexibility. For example you could subclass PlacementSpec to specify your own custom placement (ex: some database like rocksdb).

SciPioneer replied:
Can you add a comment about this? So powerful users know that they can subclass `PlacementSpec` for this purpose.

SciPioneer commented on torch/distributed/_sharding_spec/api.py:
```diff
@@ -0,0 +1,98 @@
+                2. "<worker_name>/<device>" (ex: "trainer0/cuda:0").
+    """
+
+    ShardPlacement = Union[List[Device], List[PlacementSpec]]
```
Optional: `List[Union[Device, PlacementSpec]]` can be even more flexible if you want.

SciPioneer replied:
Better to use the plural form. Otherwise I sounds more like the placement of a single shard.

SciPioneer commented on torch/distributed/_sharding_spec/api.py:
```diff
@@ -0,0 +1,98 @@
+        self._verify_dim(dim)
+        self._verify_devices(placement)
+        self._dim = dim
+        self._placement = placement
```
Ditto. `_placement` -> `_shard_placements`

SciPioneer commented on torch/distributed/_sharding_spec/api.py:
```diff
@@ -0,0 +1,98 @@
+
+Device = Union[torch.device, int, str]
+
+class PlacementSpec(ABC):
```
Why do we need a common superclass here? Will there be any major common attributes or methods?

I feel `List[DevicePlacementSpec]` should just be a data member of `ChunkShardingSpec`, and the relationship is only composition and little inheritance.

pritamdamania87 replied:
`DevicePlacementSpec` would be useful in the case where we want to shard an nn.Module. Let's say we have 3 tensors in the module and we want to place them on three different devices. `DevicePlacementSpec` can be used in conjunction with "Composite ShardingSpec" from https://github.com/pytorch/pytorch/issues/55207 to address this.

SciPioneer replied:
I see, so "sharding" here means not only intra-layer sharding, but also inter-layer sharding (where `DevicePlacementSpec`, i.e., no intra-layer sharding, can be applied).

pritamdamania committed:
Update on "Introduce ChunkShardingSpec as a model sharding specification."


Full design: https://github.com/pytorch/pytorch/issues/55207

This PR introduces ChunkShardingSpec (SingleShardingSpec in the design). Used
the name ChunkShardingSpec since it is very similar to `torch.chunk` in terms
of how a Tensor is split up and feels more clear compared to SingleShardingSpec.

Differential Revision: [D27694108](https://our.internmc.facebook.com/intern/diff/D27694108/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27694108/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Introduce ChunkShardingSpec as a model sharding specification."


Full design: https://github.com/pytorch/pytorch/issues/55207

This PR introduces ChunkShardingSpec (SingleShardingSpec in the design). Used
the name ChunkShardingSpec since it is very similar to `torch.chunk` in terms
of how a Tensor is split up and feels more clear compared to SingleShardingSpec.

Differential Revision: [D27694108](https://our.internmc.facebook.com/intern/diff/D27694108/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27694108/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Introduce ChunkShardingSpec as a model sharding specification."


Full design: https://github.com/pytorch/pytorch/issues/55207

This PR introduces ChunkShardingSpec (SingleShardingSpec in the design). Used
the name ChunkShardingSpec since it is very similar to `torch.chunk` in terms
of how a Tensor is split up and feels more clear compared to SingleShardingSpec.

Differential Revision: [D27694108](https://our.internmc.facebook.com/intern/diff/D27694108/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27694108/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Introduce ChunkShardingSpec as a model sharding specification."


Full design: https://github.com/pytorch/pytorch/issues/55207

This PR introduces ChunkShardingSpec (SingleShardingSpec in the design). Used
the name ChunkShardingSpec since it is very similar to `torch.chunk` in terms
of how a Tensor is split up and feels more clear compared to SingleShardingSpec.

Differential Revision: [D27694108](https://our.internmc.facebook.com/intern/diff/D27694108/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27694108/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Introduce ChunkShardingSpec as a model sharding specification."


Full design: https://github.com/pytorch/pytorch/issues/55207

This PR introduces ChunkShardingSpec (SingleShardingSpec in the design). Used
the name ChunkShardingSpec since it is very similar to `torch.chunk` in terms
of how a Tensor is split up and feels more clear compared to SingleShardingSpec.

Differential Revision: [D27694108](https://our.internmc.facebook.com/intern/diff/D27694108/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27694108/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Introduce ChunkShardingSpec as a model sharding specification."


Full design: https://github.com/pytorch/pytorch/issues/55207

This PR introduces ChunkShardingSpec (SingleShardingSpec in the design). Used
the name ChunkShardingSpec since it is very similar to `torch.chunk` in terms
of how a Tensor is split up and feels more clear compared to SingleShardingSpec.

Differential Revision: [D27694108](https://our.internmc.facebook.com/intern/diff/D27694108/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27694108/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Introduce ChunkShardingSpec as a model sharding specification."


Full design: https://github.com/pytorch/pytorch/issues/55207

This PR introduces ChunkShardingSpec (SingleShardingSpec in the design). Used
the name ChunkShardingSpec since it is very similar to `torch.chunk` in terms
of how a Tensor is split up and feels more clear compared to SingleShardingSpec.

Differential Revision: [D27694108](https://our.internmc.facebook.com/intern/diff/D27694108/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27694108/)!

[ghstack-poisoned]
